src/charm.py

```diff
@@ -4,27 +4,263 @@
+
+        return MySQL(
+            self._get_hostname_by_unit(self.unit.name),
+            peer_data["cluster-name"],
+            peer_data["root-password"],
+            "serverconfig",
+            peer_data["server-config-password"],
+            "clusteradmin",
+            peer_data["cluster-admin-password"],
+            self.unit.get_container("mysql"),
+        )
+
+    @property
+    def _pebble_layer(self) -> Layer:
+        """Return a layer for the pebble service."""
+        return Layer(
+            {
+                "summary": "mysqld layer",
+                "description": "pebble config layer for mysqld",
+                "services": {
+                    self._mysqld_service: {
+                        "override": "replace",
+                        "summary": "mysqld",
+                        "command": "mysqld",
+                        "startup": "enabled",
+                        "user": "mysql",
+                        "group": "mysql",
+                    }
+                },
+            }
+        )
+
+    def _get_hostname_by_unit(self, unit_name: str) -> str:
+        """Create a DNS name for a unit.
+
+        Translate juju unit name to resolvable hostname.
+
+        Args:
+            unit_name: unit name
+        Returns:
+            A string representing the hostname of the unit.
+        """
+        unit_id = unit_name.split("/")[1]
+        return f"{self.app.name}-{unit_id}.{self.app.name}-endpoints"
+
+    # =========================================================================
+    # Charm event handlers
+    # =========================================================================
+
+    def _on_config_changed(self, _) -> None:
+        """Handle the config changed event."""
+        # Only execute on unit leader
+        if not self.unit.is_leader():
+            return
+
+        # Set the cluster name in the peer relation databag if it is not already set
+        peer_data = self._peers.data[self.app]
+
+        if not peer_data.get("cluster-name"):
+            peer_data["cluster-name"] = (
+                self.config.get("cluster-name") or f"cluster-{generate_random_hash()}"
+            )
+
+    def _on_leader_elected(self, _):
+        """Handle the leader elected event.
+
+        Set config values in the peer relation databag.
+        """
+        peer_data = self._peers.data[self.app]
+
+        required_passwords = ["root-password", "server-config-password", "cluster-admin-password"]
+
+        for required_password in required_passwords:
+            if not peer_data.get(required_password):
+                logger.debug(f"Setting {required_password}")
+                password = generate_random_password(PASSWORD_LENGTH)
+                peer_data[required_password] = password
+
+        if not peer_data.get("cluster-name"):
+            peer_data["cluster-name"] = (
+                self.config.get("cluster-name") or f"cluster-{generate_random_hash()}"
```

Comment:
Same as above.

The config name is different from the name on `config.yaml`. It will not get the config value this way.
```suggestion
                self.config.get("cluster_name") or f"cluster-{generate_random_hash()}"
```

Reply:
fix on deb734a